internal/security/fileprovider/provider.go

```diff
@@ -59,6 +69,124 @@ func (p *fileTokenProvider) SetConfiguration(secretConfig secretstoreclient.Secr
+		}
+
+		// Set a meta property that consuming serices can use to automatically scope secret queries
+		createTokenParameters["meta"] = map[string]interface{}{
+			"edgex-service-name": serviceName,
+		}
+
+		// Always create a policy with this name
+		policyName := "edgex-service-" + serviceName
+
+		policyBytes, err := json.Marshal(servicePolicy)
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed encode service policy for %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		if _, err := p.vaultClient.InstallPolicy(privilegedToken, policyName, string(policyBytes)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to install policy %s: %s", policyName, err.Error()))
+			return err
+		}
+
+		var createTokenResponse interface{}
+
+		if _, err = p.vaultClient.CreateToken(privilegedToken, createTokenParameters, &createTokenResponse); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create vault token for service %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		outputTokenDir := filepath.Join(p.tokenConfig.OutputDir, serviceName)
+		outputTokenFilename := filepath.Join(outputTokenDir, p.tokenConfig.OutputFilename)
+		if err := p.fileOpener.MkdirAll(outputTokenDir, os.FileMode(0700)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create base directory path(s) %s: %s", outputTokenDir, err.Error()))
+			return err
+		}
+
+		p.logger.Info(fmt.Sprintf("creating token file %s", outputTokenFilename))
+		writeCloser, err := p.fileOpener.OpenFileWriter(outputTokenFilename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0600))
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed open token file for writing %s: %s", outputTokenFilename, err.Error()))
+			return err
+		}
+		permissionable, ok := writeCloser.(permissionable)
+		if ok {
+			if serviceConfig.FilePermissions != nil &&
+				(serviceConfig.FilePermissions).ModeOctal != nil {
+				mode, err := strconv.ParseInt(*(serviceConfig.FilePermissions).ModeOctal, 8, 32)
+				if err != nil {
+					p.logger.Error(fmt.Sprintf("invalid file mode %s: %s", *(serviceConfig.FilePermissions).ModeOctal, err.Error()))
+					return err
+				}
+				if err := permissionable.Chmod(os.FileMode(mode)); err != nil {
+					p.logger.Error(fmt.Sprintf("failed to set file mode on %s: %s", outputTokenFilename, err.Error()))
+					return err
+				}
+			}
+			if serviceConfig.FilePermissions != nil &&
+				(serviceConfig.FilePermissions).Uid != nil &&
+				(serviceConfig.FilePermissions).Gid != nil {
+				err := permissionable.Chown(*(serviceConfig.FilePermissions).Uid, *(serviceConfig.FilePermissions).Gid)
+				if err != nil {
+					p.logger.Error(fmt.Sprintf("failed to set file user/group on %s: %s", outputTokenFilename, err.Error()))
+					return err
+				}
+			}
+		}
+
+		defer writeCloser.Close()
+
+		json.NewEncoder(writeCloser).Encode(createTokenResponse) // Write resulting token
+
+		if err := writeCloser.Close(); err != nil {
```

Comment:
this is a duplicate execution of line 181 
```go
defer writeCloser.Close()
```
should consider combine that line with this one.
Suggested change:

1. Remove line 181
2. make it as defer func:
```suggestion
		defer func() {
		    if err := writeCloser.Close(); err != nil {
```

Reply:
@jim-wang-intel 

This article specifically recommends the pattern that I am using:

https://www.joeshaw.org/dont-defer-close-on-writable-files/

This file contains the vault authentication token.  I want to be absolutely sure that it is written without issues.  Moreover, I cannot `return err` inside of the deferred function to indicate to the caller than an error occurred if I defer the explicit Close().

Other suggestions?

Reply:
Could you do a defer that only closes the writeCloser if it hasn't been closed already (e.g. isn't nil).  This would catch the various error exits that might occur before you get to your explicit close (which could immediately/subsequently set writeCloser to nil to ensure the deferred close isn't executed..?

Reply:
This is possible and has been done.

Reply:
can you use err scope closure to get the error returned from inside the defer func()?  the article also mentions that as well.  like cerr := f.close() and assign the closure err = cerr? so err changes its state right before its return

Reply:
I'd strive for readability and code that's easy to understand.  My interpretation of the above is anything but.

Reply:
It is possible with a named result parameter so instead of 

```
		defer func() {
			if writeCloser != nil {
				writeCloser.Close()
			}
		}()
		// code
		if err := writeCloser.Close(); err != nil {
			p.logger.Error(fmt.Sprintf("failed to close %s: %s", outputTokenFilename, err.Error()))
			return err
		}
		writeCloser = nil
```

```
	defer func() {
		if newerr: = writeCloser.Close() {
			p.logger.Error(fmt.Sprintf("failed to close %s: %s", outputTokenFilename, err.Error()))
			err= newerr
			return
		}
	}()
	// code
```

Reply:
I would prefer to go with @michaelestrin 's suggestion as I've been burned too many times by named result arguments.

Reply:
I do not agree that the first one is more readable. It looks strange to me that if you already got  the close error ,why close again the second time no matter what?

Reply:
This PR is 2 weeks old.  I don't care which way, as long as the code gets in.  Can @jim-wang-intel and @michaelestrin  come to agreement on which way you want it done and I will do it that way.

Reply:
Your frustration is palatable and I totally understand and empathize with you.

To quote the article:

> Still, I find this pattern to be a little too magical. I dislike using named return values, and modifying the return value after the core function finishes is not intuitively clear even to experienced Go programmers.

A return value that's set in a closure isn't exactly clear or "readable."  "Too magical."  But I don't particularly care to argue the point with @jim-wang-intel and reserve my right to my opinion.

As far as I'm concerned, the current implementation looks good to me and I'll approve it.   Whether you change it or not is up to you.  :)